### ISSUE TITLE
Questions about dataset preprocessing

### ISSUE THREAD

**Issue by zijunwu02** [XuzheZ/MAPSeg#8]
Dear authors, 

Thank you for your amazing work! 

I am trying to implement the network on a CT (size = 512, 512, 400) to MRI (size = 320, 282, 302) image segmentation. However, I think (96, 96, 96) might generate an FOV too small for the image size. Therefore, I want to know whether you tried cropping or downsampling when preprocessing the cardiac dataset? What's your image input size?

Besides, I think that increasing the patch size (and keeping the source/target images the same resolution/FOV) is important. Restricted by GPU memory, increasing patch size may be difficult. Therefore, I recommend the gradient checkpoint technique `torch.utils.checkpoint` (https://pytorch.org/docs/stable/checkpoint.html) and it worked.

**Comment by zijunwu02** [comment]
Also, surprisingly, I found that if I trained the MAPSeg model from scratch repeatedly, the MPL training result is different/random for each experiment (even though the code, dataset and devices are constant). Is this resonable or is there anything wrong with my operation?

**Comment by XuzheZ** [comment]
Hello @zijunwu02 

Thank you for your questions! I am grateful for your suggestions regarding the gradient checkpoint technique. Have you already implemented that in the MAPSeg framework? Do you mind submitting a pull request to reflect that change and be a contributor to this repo? Please don't feel obligated to do that, and I am genuinely thankful for this suggestion and the information!

Sorry for the late reply about the data preprocessing, and I hope this could still be helpful. 

1. Regarding the cross-modality segmentation, for two modalities with resolution discrepancies, say you have 0.6x0.6x0.5 mm3 CT and 1.5x1.5x1.5 mm3 MRI, I will first **do the resample to match the resolution**. And the strategy _varies with your downstream task_. 
For example, if I want to segment **large organs**, I will probably downsample CT to match the MRI resolution. That downsampling gives them comparable sizes of matrices and ensures that the 96x96x96 window captures enough information. And for large organ segmentation, 1.5x1.5x1.5 mm3 is a decent resolution. 
For other **smaller targets**, or objects with more complex geometry, I will downsample CT to 1x1x1 mm3 and upsample MRI to 1x1x1 mm3, which is a level that we won't lose too much information but keep a reasonable FOV for each patch. 
If the segmentation target requires **sub-mm resolution**, it might be tricky... A brute force solution would be to **increase the patch size**, which means you also need to buy a GPU with larger memory or use some technique to calculate gradients separately. Also, it is possible to decrease the number of blocks during encoding, but again, due to the limitation of computation resources, I did not conduct such ablation studies and was unsure if the current setting is optimal. 

2. After resampling, I will roughly print the FOV of each scan; if I find a large discrepancy in the FOV of different modalities, I will do **cropping**. In some cases, the cropping is based on extra information (for example, in brain MRI, we can use a brain mask to crop the image). That is also beneficial to the training; for instance, a brain MRI with the size of 280x280x280 will have a size of 150x150x120 after cropping based on the brain mask. That's good for MAE and MPL. Therefore, if you find a way to crop unrelated information, it will be generally helpful. If there is no such automatic way to do the cropping, I usually use **center cropping** to roughly align the FOV of different modalities. 


Hmm, your finding about the inconsistent performance sounds interesting. When you said, "_trained the MAPSeg model from scratch repeatedly_", did you mean have both MAE and MPL training from scratch, using the same random seed? Using the same code, same data, same random seed, and you have trained two models using MAE and MPL, the results will be different? That is unexpected. How off the results can be? In my experiments, during the development and iteration, we had many experiments but did not observe significant differences. 
Can you check, in the saved visualization, if the order of input data is the same? Also, did you use the gradient checkpoint technique? I wonder if it will introduce randomness or mess up the training... 

Please let me know. Thank you!

Best regards,
Xuzhe

**Comment by zijunwu02** [comment]
Dear authors,

I really appreciate your detailed reply about the issue. I have just submitted a pull request regarding the gradient checkpoint technique, which saved about half of the GPU memory used in my experiment. My submitted code **only implemented the gradient checkpoint for the encoder and decoder of MAE**. More similar operations can be done inside the network (e.g. the ResNet blocks) and for MPL. Please refer to the official website (https://pytorch.org/docs/stable/checkpoint.html). But I am not sure how much the checkpoint technique might affect the performance.

About my findings of inconsistent performance, I do "_have both MAE and MPL training from scratch, using the same random seed_". To be more specific, I trained the MAPSeg (both MAE & MPL) repeatedly with exactly same codes, configuration files (same random seeds), and datasets (MAE was trained with 16 target scans and 16 source scans; MPL was trained with 16 source scans and 16 target scans, and validated with 4 source scans). MAE was trained 75 epochs (scheduled at last 25 epochs) and MPL was trained 75 epochs (warm-up at first 25 epochs, all scheduled). Gradient checkpoint wasn't used.

I found that MAE performances are consistent (print exactly same losses). The MPL performances are NOT consistent. Checking the visualization file, I found the patch/scan, label, masked_map are all constant across different experiments, only the predictions are different. I guess that it may be caused by the `dropout` in the MPL network? And I think **the differences might be reduced will larger scale datasets**.

Thanks again for your sincere reply!

Best regards,
Zijun

**Comment by yuaoli** [comment]
Dear author:
Thank you so much for your excellent work!
When I am replicating your work, I find that the README does not mention how the model weights are obtained as teacher models in MPL, where should I download them or which data set should I train on?
